src/repr/src/fixed_length.rs

```diff
@@ -0,0 +1,752 @@
+        let iter = row.iter();
+        Self::from_datum_iter(iter, types)
+    }
+
+    /// This implementation reads the datums in `row`, validates their types
+    /// against the schema provided by `types`, and returns a fixed-length
+    /// N-byte value as a result.
+    ///
+    /// Presently, the implementation panics if more than one `ColumnType` or
+    /// a type larger than what fits in an N-byte representation is found in `types`.
+    /// Additionally, the implementation panics if decoding the row results in an
+    /// unexpected type.
+    fn from_datum_iter<'a, I, D>(iter: I, types: &[ColumnType]) -> Self
+    where
+        I: IntoIterator<Item = D>,
+        D: Borrow<Datum<'a>>,
+    {
+        // TODO(vmarcos): In the future, the implementation should be extended to
+        // panic if a multi-column `row` has more data than fits in an N-byte tagless
+        // representation with one byte reserved for nullability, or if the types of
+        // the datums do not match with `types`.
+        assert!(N > 1);
+        assert!(types.len() == 1);
+        let typ = &types[0];
+        let mut datum_iter = iter.into_iter();
+        let next_datum = datum_iter
+            .next()
+            .expect("At least one Datum expected in fixed-length encoding");
+        let next_datum = *next_datum.borrow();
+        let mut result = BytesN { data: [0; N] };
+
+        // Firstly, try to encode Datum::Null.
+        if let Datum::Null = next_datum {
+            result.data[0] = 1u8;
+        } else {
+            // Non-null datums should fit the remaining bytes.
+            let mut offset = 1;
+            let typ_len = fixed_length(&typ.scalar_type).unwrap_or_else(|| {
+                panic!(
+                    "Type {:?} without fixed-length for encoding",
+                    typ.scalar_type
+                )
+            });
+            if typ_len > N - offset {
+                panic!(
+                    "Type {:?} too wide for fixed-length {} encoding",
+                    typ.scalar_type, N
+                );
+            }
+            match typ.scalar_type {
+                ScalarType::Bool => {
+                    result.data[offset] = match next_datum {
+                        Datum::True => 1u8,
+                        Datum::False => 0u8,
+                        _ => panic!(
+                            "Type {:?} expected, but row contained {:?}",
+                            typ.scalar_type, next_datum
+                        ),
+                    };
+                }
+                ScalarType::Int16 => {
+                    let Datum::Int16(n) = next_datum else {
+                        panic!(
+                            "Type {:?} expected, but row contained {:?}",
+                            typ.scalar_type, next_datum
+                        )
+                    };
+                    result.data[offset..offset + size_of::<i16>()]
```

Comment:
similarly, all these could be simplified with the endio crate and `std::io::Write`.

Reply:
In addition to https://github.com/MaterializeInc/materialize/pull/21968#discussion_r1356535883, what I wanted to get was just access to `memcpy`, which apparently `copy_from_slice` is a blessed way of getting: https://users.rust-lang.org/t/how-to-memcpy-bytes-in-stable-rust/2740. The implementation of `read_byte_array` actually just calls `copy_from_slice` too.

Reply:
The [documentation](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) of `copy_from_slice` says that the function will panic if "the two slices have different lengths." So I think that we need to keep the upper boundary there, unfortunately. 

The code would have indeed been a bit less noisy without these, and they aren't strictly necessary as an assert because we perform bounds checking on the `fixed_length(..)` of the type being decoded earlier in the function.

Reply:
Ah, yes I overlooked that!

You could write a helper function to hid the boilerplate, e.g.:

```rust
fn copy_into(dst: &mut [u8], src: &[u8]) {
    let dst = &mut dst[..src.len()];
    dst.copy_from_slice(src);
}
```

But I'm not sure if that would still be useful once you want to pack multiple Datums into one `BytesN`.

Reply:
Nice suggestion; I adapted it a bit, and it simplified the encoding code a lot.